src/request.rs

```diff
@@ -238,27 +267,176 @@ impl RequestBuilder {
     }
 }
 
+impl MultipartRequestBuilder {
+    /// Constructs a new request.
+    pub fn new(request_builder: RequestBuilder) -> MultipartRequestBuilder {
+        MultipartRequestBuilder { request_builder: request_builder, params: None, files: None }
+    }
+
+    /// Add a `Header` to this Request.
+    ///
+    /// ```rust
+    /// # use reqwest::Error;
+    /// #
+    /// # fn run() -> Result<(), Error> {
+    /// use reqwest::header::UserAgent;
+    /// let client = reqwest::Client::new()?;
+    ///
+    /// let res = client.get("https://www.rust-lang.org")?
+    ///     .header(UserAgent("foo".to_string()))
+    ///     .send()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn header<H>(&mut self, header: H) -> &mut MultipartRequestBuilder
+        where H: ::header::Header + ::header::HeaderFormat
+    {
+        self.request_builder.request_mut().headers.set(header);
+        self
+    }
+
+    /// Add a set of Headers to the existing ones on this Request.
+    ///
+    /// The headers will be merged in to any already set.
+    pub fn headers(&mut self, headers: ::header::Headers) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().headers.extend(headers.iter());
+        self
+    }
+
+    /// Enable HTTP basic authentication.
+    pub fn basic_auth<U, P>(&mut self, username: U, password: Option<P>) -> &mut MultipartRequestBuilder
+        where U: Into<String>,
+              P: Into<String>
+    {
+        self.request_builder.header(::header::Authorization(::header::Basic {
+                                                username: username.into(),
+                                                password: password.map(|p| p.into()),
+                                            }));
+        self
+    }
+
+    /// Set the request body.
+    pub fn body<T: Into<Body>>(&mut self, body: T) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().body = Some(body.into());
+        self
+    }
+
+    /// Add parameters to the multipart/form-data `Request`.
+    pub fn params(&mut self, params: Params) -> &mut MultipartRequestBuilder {
+        self.params = Some(params);
+        self
+    }
+
+    /// Add files to the multipart/form-data `Request`.
+    pub fn files(&mut self, files: Files) -> &mut MultipartRequestBuilder {
```

Comment:
I was considering doing that, but that's a lot of chaining, to me it seemed like making a `Vec<File>` was the nicer way, but if you want to go a full builder pattern it would make more sense. 

We could make the `Mime` an `Option<Mime>` and use [mime_guess](https://crates.io/crates/mime_guess) internally for when `Mime` is `None`.